subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c

```diff
@@ -0,0 +1,928 @@
+
+	enum nrf_cloud_fota_type type;
+	char *id;
+	char *host;
+	char *path;
+	int file_size;
+
+	enum nrf_cloud_fota_error error;
+	int dl_progress;
+	/* tracking for CONFIG_NRF_CLOUD_FOTA_PROGRESS_PCT_INCREMENT */
+	int sent_dl_progress;
+};
+
+struct settings_fota_job {
+	enum fota_validate_status validate;
+	enum nrf_cloud_fota_type type;
+	char id[JOB_ID_STRING_SIZE];
+};
+
+static void http_fota_handler(const struct fota_download_evt *evt);
+static void send_event(const enum nrf_cloud_fota_evt_id id,
+		       const struct nrf_cloud_fota_job * const job);
+
+static void cleanup_job(struct nrf_cloud_fota_job *const job);
+static int start_job(struct nrf_cloud_fota_job *const job);
+static int send_job_update(struct nrf_cloud_fota_job *const job);
+static int publish(const struct mqtt_publish_param *const pub);
+static bool is_fota_active(void);
+static int save_validate_status(const char *const job_id,
+			   const enum nrf_cloud_fota_type job_type,
+			   const enum fota_validate_status status);
+static int fota_settings_set(const char *key, size_t len_rd,
+			     settings_read_cb read_cb, void *cb_arg);
+
+static struct mqtt_client *client_mqtt;
+static nrf_cloud_fota_callback_t event_cb;
+
+static struct mqtt_topic topic_rcv = { .qos = MQTT_QOS_1_AT_LEAST_ONCE };
+static struct mqtt_topic topic_update = { .qos = MQTT_QOS_1_AT_LEAST_ONCE };
+static struct mqtt_topic topic_req = { .qos = MQTT_QOS_1_AT_LEAST_ONCE };
+static const struct mqtt_subscription_list sub_list = {
+	.list = &topic_rcv,
+	.list_count = 1,
+	.message_id = NRF_CLOUD_FOTA_SUBSCRIBE_ID,
+};
+
+static enum fota_download_evt_id last_fota_dl_evt = FOTA_DOWNLOAD_EVT_ERROR;
+static struct nrf_cloud_fota_job current_fota;
+static struct settings_fota_job saved_job = { .type = NRF_FOTA_TYPE__INVALID };
+static bool initialized;
+
+#define SETTINGS_KEY_FOTA "fota"
+#define SETTINGS_FULL_FOTA NRF_CLOUD_SETTINGS_NAME \
+			   "/" \
+			   SETTINGS_KEY_FOTA
+#define SETTINGS_FOTA_JOB "job"
+#define SETTINGS_FULL_FOTA_JOB SETTINGS_FULL_FOTA \
+			       "/" \
+			       SETTINGS_FOTA_JOB
+
+SETTINGS_STATIC_HANDLER_DEFINE(fota, SETTINGS_FULL_FOTA, NULL,
+			       fota_settings_set, NULL, NULL);
+
+static int fota_settings_set(const char *key, size_t len_rd,
+			     settings_read_cb read_cb, void *cb_arg)
+{
+	if (!key) {
+		LOG_DBG("Key is NULL");
+		return -EINVAL;
+	}
+
+	LOG_DBG("Settings key: %s, size: %d", log_strdup(key), len_rd);
+
+	if (!strncmp(key, SETTINGS_FOTA_JOB, strlen(SETTINGS_FOTA_JOB)) &&
+	    (len_rd == sizeof(saved_job))) {
+		if (read_cb(cb_arg, (void *)&saved_job, len_rd) == len_rd) {
+			LOG_DBG("Saved job: %s, type: %d, validate: %d",
+				log_strdup(saved_job.id), saved_job.type,
+				saved_job.validate);
+			return 0;
+		}
+	}
+	return -ENOTSUP;
+}
+
+enum fota_validate_status get_modem_update_status(void)
+{
+	enum fota_validate_status ret = NRF_FOTA_VALIDATE_UNKNOWN;
+
+#if defined(CONFIG_BSD_LIBRARY)
+	int modem_dfu_res = bsdlib_get_init_ret();
```

Comment:
Not the correct place for the change  just FYI:
You can now deinitialize and initialize the modem without having to power-cycle the chip to perform a modem update.

I think the sequence is
```
bsd_shutdown();
int modem_dfu_res;
check_result:
modem_dfu_res = bsd_init();
/* Handle return values relating to modem firmware update */
	switch (modem_dfu_res) {
	case MODEM_DFU_RESULT_OK:
		LOG_DBG("Modem FOTA OK");
		ret = NRF_FOTA_VALIDATE_PASS;
                 bsd_shutdown();
                 goto check_result;
	case MODEM_DFU_RESULT_UUID_ERROR:
	case MODEM_DFU_RESULT_AUTH_ERROR:
	case MODEM_DFU_RESULT_HARDWARE_ERROR:
	case MODEM_DFU_RESULT_INTERNAL_ERROR:
		LOG_ERR("Modem FOTA error: %d", modem_dfu_res);
		ret = NRF_FOTA_VALIDATE_FAIL;
		break;
	default:
		LOG_DBG("Modem FOTA result unknown: %d", modem_dfu_res);
		break;
	}

Reply:
ah nice, good to know.

Reply:
I think it's bsdlib specific

Reply:
i need to rebase and retest, initially this did not work as expected.